src/datachain/lib/listing.py

```diff
@@ -167,12 +155,20 @@ def get_listing(
     client = Client.get_client(uri, cache, **client_config)
     telemetry.log_param("client", client.PREFIX)
     if not isinstance(uri, str):
-        uri = str(uri)
+        uri = os.fspath(uri)
 
     # we don't want to use cached dataset (e.g. for a single file listing)
-    if not glob.has_magic(uri) and not uri.endswith("/") and _isfile(client, uri):
-        storage_uri, path = Client.parse_url(uri)
-        return None, f"{storage_uri}/{path.lstrip('/')}", path, False
+    isfile = _reraise_as_client_error()(fsutils.isfile)
+    if not glob.has_magic(uri) and isfile(client.fs, uri):
+        if uri.endswith(client.fs.sep):
+            # should be fixed upstream in fsspec
+            # boils down to https://github.com/fsspec/filesystem_spec/pull/1567#issuecomment-2563160414
+            # fsspec removes the trailing slash and returns a file, that's why we are
+            # are not getting an error here
+            raise NotADirectoryError(errno.ENOTDIR, os.strerror(errno.ENOTDIR), uri)
```

Comment:
cc @shcheklein. I changed this to raise `NotADirectoryError` when the given path is a path to a file and has a trailing slash.

```console
$ datachain ls pyproject.toml/
Error: [Errno 20] Not a directory: 'pyproject.toml/'
```

I think it's better to fail eagerly here.

Reply:
I think we could also move this logic inside `isfile`, but that won't allow us to raise `NotADirectoryError`.

Reply:
It looks like `azure` and `gs` raise `FileNotFoundError` in this case.